client/state/terms/selectors.js

```diff
@@ -33,27 +37,97 @@ export function isRequestingTermsForQuery( state, siteId, taxonomy, query ) {
  * @param  {Object}  state    Global state tree
  * @param  {Number}  siteId   Site ID
  * @param  {String}  taxonomy Taxonomy slug
- * @param  {Object}  query    Post query object
- * @return {?Array}           Posts for the post query
+ * @param  {Object}  query    Terms query object
+ * @return {?Array}           Terms for the query
  */
-export function getTermsForQuery( state, siteId, taxonomy, query ) {
-	const serializedQuery = getSerializedTermsQuery( query );
-	const queryResults = get( state.terms.queries, [ siteId, taxonomy, serializedQuery ] );
-	if ( ! queryResults ) {
-		return null;
+export const getTermsForQuery = createSelector(
+	( state, siteId, taxonomy, query ) => {
+		const serializedQuery = getSerializedTermsQuery( query );
+		const queryResults = get( state.terms.queries, [ siteId, taxonomy, serializedQuery ] );
+		if ( ! queryResults ) {
+			return null;
+		}
+
+		return queryResults.reduce( ( memo, termId ) => {
+			const term = get( state.terms, [ 'items', siteId, taxonomy, termId ] );
+			if ( term ) {
+				memo.push( term );
+			}
+
+			return memo;
+		}, [] );
+	},
+	( state, siteId, taxonomy ) => getTerms( state, siteId, taxonomy ),
+	( state, siteId, taxonomy, query ) => {
+		const serializedQuery = getSerializedTermsQuery( query );
+		return [ siteId, taxonomy, serializedQuery ].join();
 	}
+);
 
-	return queryResults.reduce( ( memo, termId ) => {
-		const term = get( state.terms, [ 'items', siteId, taxonomy, termId ] );
-		if ( term ) {
-			memo.push( term );
-		}
+/**
+ * Returns an array of terms for the taxonomy query, including all known
+ * queried pages, or null if the number of pages is unknown.
+ *
+ * @param  {Object}  state    Global state tree
+ * @param  {Number}  siteId   Site ID
+ * @param  {String}  taxonomy Taxonomy slug
+ * @param  {Object}  query    Terms query object
+ * @return {?Array}           Terms for the query
+ */
+export function getTermsForQueryIgnoringPage( state, siteId, taxonomy, query ) {
+	const lastPage = getTermsLastPageForQuery( state, siteId, taxonomy, query );
+	if ( null === lastPage ) {
+		return lastPage;
+	}
 
-		return memo;
+	return range( 1, lastPage + 1 ).reduce( ( memo, page ) => {
+		const termsQuery = Object.assign( {}, query, { page } );
+		return memo.concat( getTermsForQuery( state, siteId, taxonomy, termsQuery ) || [] );
 	}, [] );
 }
 
 /**
+ * Returns an hierarchical array of terms for the taxonomies query, or null if no terms have been
+ * received.
+ *
+ * @param  {Object}  state    Global state tree
+ * @param  {Number}  siteId   Site ID
+ * @param  {String}  taxonomy Taxonomy slug
+ * @param  {Object}  query    Term query object
+ * @return {?Array}           Terms for the query
+ */
+export const getTermsHierarchyForQueryIgnoringPage = createSelector(
```

Comment:
>For the initial usage of porting `<CategorySelector />` to use `<TermsSelector />` (#5366).  We show the tree of categories currently in both the sidebar of the editor and the menu builder

Whoops, you're right. I loaded up the editor to check whether we were showing hierarchy, and saw for my test site that we weren't... but that was only because I didn't have any nested categories 😉